src/connection.rs

```diff
@@ -172,13 +181,47 @@ impl BrokerConnector {
     ///
     /// If `broker_override` is provided this will request metadata from that specific broker.
     ///
-    /// Note: overriding the broker prevents automatic handling of connection-invalidating errors,
-    /// which will instead be returned to the caller
+    /// Note: overriding the broker prevents automatic handling of connection-invalidating errors, which will instead be
+    /// returned to the caller
+    ///
+    /// # Cache
+    ///
+    /// The caller can request a cached metadata entry by specifying `use_cache: true` in the call. If no cached entry
+    /// is available the request will be dispatched to an arbitrary broker.
+    ///
+    /// Using a cached entry is only valid if:
+    ///  * An arbitrary broker was to be queried
+    ///  * The caller can tolerate, or validate, a potentially infinitely stale cached entry
+    ///
+    /// # Panics
+    ///
+    /// It is invalid to request metadata from a specific broker (by specifying `broker_override`), and request a cached
+    /// entry - doing so will panic.
     pub async fn request_metadata(
         &self,
         broker_override: Option<BrokerConnection>,
         topics: Option<Vec<String>>,
+        use_cache: bool,
     ) -> Result<MetadataResponse> {
+        // Return a cached metadata response as an optimisation to prevent
+        // multiple successive metadata queries for the same topic across
+        // multiple PartitionClient instances.
+        //
+        // NOTE: no serialisation / locking is imposed over the cache during
+        // this call  - multiple parallel calls to this function will still
+        // perform multiple requests until the cache is populated. However, the
+        // Client initialises this cache at construction time, so unless
+        // invalidated, there will always be a cached entry available.
+        if use_cache {
+            assert!(
+                broker_override.is_none(),
+                "cannot ask for cached metadata from specific broker"
+            );
```

Comment:
The cache maintains metadata acquired from an arbitrary broker, so it is an error to request a cached entry from a specific broker.

Reply:
Maybe instead of using two bi-states (`Option` and `bool`) with one invalid combination use an enum to encode the tri-state ("use cache", "fetch fresh from arbitrary broker", "fetch fresh from specific broker"). The same goes for `get_leader`.